### PR TITLE
extend init to accept rasr binaries

### PR DESCRIPTION
**Pull request by christophmluscher**
the idea is to set the rasr binaries at the beginning. at the moment this still uses `gs.RASR_*` but we could remove this. it would be cleaner but backward comp?!

### CODE REVIEW

**Comment by christophmluscher** [comment]
we might want to wait for merging until Nicks PR (in preparation) is merged first..

File: common/setups/rasr/rasr_system.py [hunk 1]
```diff
@@ -49,9 +49,27 @@ class RasrSystem(meta.System):
...
+        self,
+        rasr_root: Optional[tk.Path] = None,
+        rasr_arch: Optional[str] = None,
+    ):
         super().__init__()
 
+        if rasr_root is not None:
+            self.rasr_root = rasr_root
+        elif hasattr(gs, "RASR_ROOT") and gs.RASR_ROOT is not None:
+            self.rasr_root = gs.RASR_ROOT
```
[the remark this answers is not in the document]
Comment (christophmluscher) [reply common/setups/rasr/rasr_system.py:62]: I would remove `gs.RASR_*` completely

**Comment by michelwi** [comment]
> it would be cleaner but backward comp?!

yes, I also agree to remove it completely. If a user wants to use a specific RASR version from a certain path in the FS, then he could just give it as a Path object to this function.

But now the argument `rasr_root` should be required and not optional. and `rasr_arch` can default to "linux-x86_64-standard" and then you can remove the if-else-logic when you set it later.

File: common/setups/rasr/gmm_system.py [hunk 2]
```diff
@@ -88,8 +88,12 @@ class GmmSystem(RasrSystem):
     corpus -> Path
     """
 
-    def __init__(self):
-        super().__init__()
+    def __init__(
+        self,
+        rasr_root: Optional[tk.Path] = None,
```
Comment (JackTemaki) [inline common/setups/rasr/gmm_system.py:93]:
```suggestion
        rasr_binary_path: Optional[tk.Path] = None,
```
We are not not passing the root, but a path to a folder directly containing the binaries, as they can be independent from the repository.

File: common/setups/rasr/gmm_system.py [hunk 3]
```diff
@@ -88,8 +88,12 @@ class GmmSystem(RasrSystem):
...
+        rasr_arch: Optional[str] = None,
+    ):
+        super().__init__(rasr_root=rasr_root, rasr_arch=rasr_arch)
```
Comment (JackTemaki) [inline common/setups/rasr/gmm_system.py:96]:
```suggestion
        super().__init__(rasr_binary_path=rasr_binary_path, rasr_arch=rasr_arch)
```

File: common/setups/rasr/rasr_system.py [hunk 4]
```diff
@@ -49,9 +49,16 @@ class RasrSystem(meta.System):
     corpus -> Path
     """
 
-    def __init__(self):
+    def __init__(
+        self,
+        rasr_root: Optional[tk.Path] = None,
```
Comment (JackTemaki) [inline common/setups/rasr/rasr_system.py:54]:
```suggestion
        rasr_binary_path: Optional[tk.Path] = None,
```

File: common/setups/rasr/rasr_system.py [hunk 5]
```diff
@@ -49,9 +49,16 @@ class RasrSystem(meta.System):
...
+        rasr_arch: Optional[str] = None,
+    ):
         super().__init__()
 
+        self.rasr_root = rasr_root
```
Comment (JackTemaki) [inline common/setups/rasr/rasr_system.py:59]:
```suggestion
        self.rasr_binary_path = rasr_binary_path
```

File: common/setups/rasr/rasr_system.py [hunk 6]
```diff
@@ -102,6 +109,14 @@ def _init_corpus(self, corpus_key: str):
             concurrent=self.concurrent[corpus_key],
             segment_path=segm_corpus_job.out_segment_path,
         )
+
+        if self.rasr_root is not None and self.rasr_arch is not None:
```
Comment (JackTemaki) [inline common/setups/rasr/rasr_system.py:113]:
```suggestion
        if self.rasr_binary_path is not None and self.rasr_arch is not None:
```

File: common/setups/rasr/rasr_system.py [hunk 7]
```diff
@@ -102,6 +109,14 @@ def _init_corpus(self, corpus_key: str):
...
+            self.crp[corpus_key].set_executables(
+                rasr_root=self.rasr_root, rasr_arch=self.rasr_arch
```
Comment (JackTemaki) [inline common/setups/rasr/rasr_system.py:115]:
```suggestion
                rasr_root=self.rasr_binary_path, rasr_arch=self.rasr_arch
```

File: common/setups/rasr/rasr_system.py [hunk 8]
```diff
@@ -102,6 +109,14 @@ def _init_corpus(self, corpus_key: str):
...
+            )
+        elif self.rasr_root is not None and self.rasr_arch is None:
```
Comment (JackTemaki) [inline common/setups/rasr/rasr_system.py:117]:
```suggestion
        elif self.rasr_binary_path is not None and self.rasr_arch is None:
```

File: common/setups/rasr/rasr_system.py [hunk 9]
```diff
@@ -102,6 +109,14 @@ def _init_corpus(self, corpus_key: str):
...
+            self.crp[corpus_key].set_executables(rasr_root=self.rasr_root)
```
Comment (JackTemaki) [inline common/setups/rasr/rasr_system.py:118]:
```suggestion
            self.crp[corpus_key].set_executables(rasr_binary_path=self.rasr_binary_path)
```

**Review by JackTemaki — CHANGES_REQUESTED** [review]

File: common/setups/rasr/rasr_system.py [hunk 10]
```diff
@@ -102,6 +109,14 @@ def _init_corpus(self, corpus_key: str):
             concurrent=self.concurrent[corpus_key],
             segment_path=segm_corpus_job.out_segment_path,
         )
+
+        if self.rasr_binary_path is not None and self.rasr_arch is not None:
+            self.crp[corpus_key].set_executables(
+                rasr_root=self.rasr_binary_path, rasr_arch=self.rasr_arch
```
Comment (JackTemaki) [inline common/setups/rasr/rasr_system.py:115]:
```suggestion
                rasr_binary_path=self.rasr_binary_path, rasr_arch=self.rasr_arch
```

**Review by JackTemaki — APPROVED** [review]

**Comment by michelwi** [comment]
I still stand to my previous comment/suggestion:

> the argument rasr_root should be required and not optional. and rasr_arch can default to "linux-x86_64-standard" and then you can remove the if-else-logic when you set it later.

And I think a doc string would be helpful to avoid the confusion that chris/me originally had

> We are not not passing the root, but a path to a folder directly containing the binaries, as they can be independent from the repository.

But otherwise I am ok with the current state of the code.

File: common/setups/rasr/rasr_system.py [hunk 11]
```diff
@@ -102,6 +114,14 @@ def _init_corpus(self, corpus_key: str):
...
             segment_path=segm_corpus_job.out_segment_path,
         )
+
+        if self.rasr_binary_path is not None and self.rasr_arch is not None:
+            self.crp[corpus_key].set_executables(
+                rasr_binary_path=self.rasr_binary_path, rasr_arch=self.rasr_arch
+            )
+        elif self.rasr_binary_path is not None and self.rasr_arch is None:
+            self.crp[corpus_key].set_executables(rasr_binary_path=self.rasr_binary_path)
+
```
Comment (michelwi) [inline common/setups/rasr/rasr_system.py:124]:
```suggestion
        self.crp[corpus_key].set_executables(
            rasr_binary_path=self.rasr_binary_path, rasr_arch=self.rasr_arch
        )
```
If the defaults are not None any more, we don't need to test for it.

**Review by michelwi — APPROVED** [review]

**Pull request merged**